Add more descriptive error details to better diagnose URLPattern test failures

jelee53 opened the pull request:
#### 55c279735d5117f113f22590f07b9f075bef3f1b
<pre>
Add more descriptive error details to better diagnose URLPattern test failures
<a href="https://bugs.webkit.org/show_bug.cgi?id=283384">https://bugs.webkit.org/show_bug.cgi?id=283384</a>
<a href="https://rdar.apple.com/140233854">rdar://140233854</a>

Reviewed by Sihui Liu.

It would be helpful to know why Tokenizer::processTokenizingError occurred and thus a reason / string should be forwarded. OpenSource/LayoutTests/fast/urlpattern/urlpattern-component.html also should better track which test case is being executed.

* LayoutTests/fast/urlpattern/urlpattern-component-expected.txt:
* LayoutTests/fast/urlpattern/urlpattern-component.html:
* Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
(WebCore::URLPatternUtilities::Tokenizer::processTokenizingError):
(WebCore::URLPatternUtilities::Tokenizer::tokenize):
* Source/WebCore/Modules/url-pattern/URLPatternTokenizer.h:

Canonical link: <a href="https://commits.webkit.org/286836@main">https://commits.webkit.org/286836@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aaee907fa0041b423c8c364e425193d1962e7ca7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77236 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56271 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30151 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81801 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28518 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79353 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65419 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4567 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60503 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/18547 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80303 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50476 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66289 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40803 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47879 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23787 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26841 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68989 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24122 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83224 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4616 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/3114 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68778 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4772 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66262 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68035 "Found 1 new API test failure: /TestWebKit:WebKit.InjectedBundleInitializationUserDataCallbackWins (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16991 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12027 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10124 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4563 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/7378 "Failed to build and analyze WebKit") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4582 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8017 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6341 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.h:
```diff
@@ -63,7 +63,7 @@ class Tokenizer {
     void addToken(TokenType currentType);
 
-    ExceptionOr<void> processTokenizingError(size_t nextPosition, size_t valuePosition);
+    ExceptionOr<void> processTokenizingError(size_t nextPosition, size_t valuePosition, const String&&);
```
Either `const String&` or `String&&`

jelee53 replied:
Oops my bad! Thanks for catching this!

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
```diff
@@ -132,7 +132,7 @@ ExceptionOr<Vector<Token>> Tokenizer::tokenize()
         if (m_codepoint == '\\') {
             if (m_index == m_input.length() - 1) {
-                maybeException = processTokenizingError(m_nextIndex, m_index);
+                maybeException = processTokenizingError(m_nextIndex, m_index, "No character provided after escape."_s);
```
Nit: `is provided`

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
```diff
@@ -187,28 +187,28 @@ ExceptionOr<Vector<Token>> Tokenizer::tokenize()
 
                 if (regexPosition == regexStart && m_codepoint == '?') {
-                    maybeException = processTokenizingError(regexStart, m_index);
+                    maybeException = processTokenizingError(regexStart, m_index, "Regex can not start with modifier."_s);
```
Nit: cannot

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
```diff
@@ -187,28 +187,28 @@ ExceptionOr<Vector<Token>> Tokenizer::tokenize()
                 if (m_codepoint == '\\') {
                     if (m_index == m_input.length() - 1) {
-                        maybeException = processTokenizingError(regexStart, m_index);
+                        maybeException = processTokenizingError(regexStart, m_index, "No character provided after escape."_s);
```
Nit: is provided

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
```diff
@@ -230,7 +230,7 @@ ExceptionOr<Vector<Token>> Tokenizer::tokenize()
 
                     if (m_index == m_input.length() - 1) {
-                        maybeException = processTokenizingError(regexStart, m_index);
+                        maybeException = processTokenizingError(regexStart, m_index, "No closing token provided by end of string."_s);
```
Nit: `is provided`

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
```diff
@@ -239,7 +239,7 @@ ExceptionOr<Vector<Token>> Tokenizer::tokenize()
 
                     if (m_codepoint != '?') {
-                        maybeException = processTokenizingError(regexStart, m_index);
+                        maybeException = processTokenizingError(regexStart, m_index, "Required OtherModifier token not provided in regex."_s);
```
Nit: `is not provided`

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
```diff
@@ -91,10 +91,10 @@ void Tokenizer::addToken(TokenType currentType)
 {
     if (m_policy == TokenizePolicy::Strict)
-        return Exception { ExceptionCode::TypeError, "Invalid token provided to URLPattern tokenizer with strict tokenize policy."_s };
+        return Exception { ExceptionCode::TypeError, makeString("Invalid token provided to URLPattern tokenizer with strict tokenize policy -- "_s, callerErrorInfo) };
```
Also we may directly print info here since we will only throw error if the policy is strict (so no need to emphasize error is caused by strict policy):
`return Exception { ExceptionCode::TypeError, WTFMove(errorMessage) };`

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
```diff
@@ -91,10 +91,10 @@ void Tokenizer::addToken(TokenType currentType)
 {
     if (m_policy == TokenizePolicy::Strict)
-        return Exception { ExceptionCode::TypeError, "Invalid token provided to URLPattern tokenizer with strict tokenize policy."_s };
+        return Exception { ExceptionCode::TypeError, WTFMove(callerErrorInfo) };
```
We should only use `WTFMove()` if parameter is `String&&`...

jelee53 committed:
Add more descriptive error details to better diagnose URLPattern test failures
https://bugs.webkit.org/show_bug.cgi?id=283384
rdar://140233854

Reviewed by Sihui Liu.

It would be helpful to know why Tokenizer::processTokenizingError occurred and thus a reason / string should be forwarded. OpenSource/LayoutTests/fast/urlpattern/urlpattern-component.html also should better track which test case is being executed.

* LayoutTests/fast/urlpattern/urlpattern-component-expected.txt:
* LayoutTests/fast/urlpattern/urlpattern-component.html:
* Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
(WebCore::URLPatternUtilities::Tokenizer::processTokenizingError):
(WebCore::URLPatternUtilities::Tokenizer::tokenize):
* Source/WebCore/Modules/url-pattern/URLPatternTokenizer.h:

Canonical link: https://commits.webkit.org/286836@main
